src/trace/xray-service.ts

```diff
@@ -0,0 +1,204 @@
+  private header?: string;
+  private context?: XrayTraceHeader;
+  private readonly subsegmentName = "datadog-metadata";
+  private readonly subsegmentNamespace = "datadog";
+  private readonly baggageSubsegmentKey = "root_span_metadata";
+  private readonly subsegmentKey = "trace";
+  private readonly lambdaFunctionTagsKey = "lambda_function_tags";
+
+  constructor() {
+    this.header = process.env[AMZN_TRACE_ID_ENV_VAR];
+    if (this.header === undefined) {
+      logDebug("Couldn't read Xray trace header from env");
+    }
+
+    this.context = this.parseTraceContextHeader();
+    if (this.context === undefined) {
+      logDebug("Couldn't parse Xray trace header from env");
+    }
+  }
+
+  public addLambdaTriggerTags(triggerTags: { [key: string]: string }) {
+    this.add(this.lambdaFunctionTagsKey, triggerTags);
+  }
+
+  public addStepFunctionContext(context: StepFunctionContext) {
+    this.add(this.baggageSubsegmentKey, context);
+  }
+
+  public addMetadata(metadata: Record<string, any>) {
+    this.add(this.subsegmentKey, metadata);
+  }
+
+  private add(key: string, metadata: Record<string, any>) {
+    const subsegment = this.generateSubsegment(key, metadata);
+
+    if (subsegment === undefined) return;
+
+    this.sendSubsegment(subsegment);
+  }
+
+  private generateSubsegment(key: string, metadata: Record<string, any>) {
+    if (this.context === undefined) return;
+
+    const sampled = this.convertToSampleMode(parseInt(this.context.sampled, 10));
+    if (sampled === SampleMode.USER_REJECT || sampled === SampleMode.AUTO_REJECT) {
+      logDebug("Discarding Xray metadata subsegment due to sampling");
+      return;
+    }
+
+    const milliseconds = Date.now() * 0.001;
+
+    return JSON.stringify({
+      id: randomBytes(8).toString("hex"),
+      trace_id: this.context.traceId,
+      parent_id: this.context.parentId,
+      name: this.subsegmentName,
+      start_time: milliseconds,
+      end_time: milliseconds,
+      type: "subsegment",
+      metadata: {
+        [this.subsegmentNamespace]: {
+          [key]: metadata,
+        },
+      },
+    });
+  }
+
+  private parseTraceContextHeader(): XrayTraceHeader | undefined {
+    if (this.header === undefined) return;
+
+    // Example: Root=1-5e272390-8c398be037738dc042009320;Parent=94ae789b969f1cc5;Sampled=1
+    logDebug(`Reading Xray trace context from env var ${this.header}`);
+    const [root, parent, _sampled] = this.header.split(";");
+    if (parent === undefined || _sampled === undefined) return;
+
+    const [, traceId] = root.split("=");
+    const [, parentId] = parent.split("=");
+    const [, sampled] = _sampled.split("=");
+    if (traceId === undefined || parentId === undefined || sampled === undefined) return;
+
+    return {
+      traceId,
+      parentId,
+      sampled,
+    };
+  }
+
+  private convertToSampleMode(xraySampled: number): SampleMode {
+    return xraySampled === 1 ? SampleMode.USER_KEEP : SampleMode.USER_REJECT;
```

Comment:
We should check to see if this is the correct mapping. I set the mode to `AUTO_KEEP` when the sampled value is `1` in the extension. I believe I pulled that logic from the java tracer.

Reply:
Is that done in Xray? If so, then we should revise these values, either way, this is 4 year old code that I'm not aware which value is the right one!

Reply:
Sounds good. Anything that's just a copy from the previous code, you're right we should just leave it how it was.